### ISSUE TITLE
How should I add contrib modules to my composer build?

### ISSUE THREAD

**Issue by laurencefass** [geerlingguy/drupal-vm#1263]
### Issue Type

 - Documentation Report

### Your Environment

```
Vagrant 1.9.3
VBox 5.1.18
Ansible 2.2.1.0
```

### Your OS

Debian

### Summary

Im using the config.yml "drupal_build_composer: true" option based on the last issue I raised.

Out of the box DrupalVM adds and enables the devel module. I would like to be able to add contrib more modules. Whats the correct procedure to add and enable contrib modules so I dont break anything? 

thanks

**Comment by laurencefass** [comment]
my initial best guess...
drupal_composer_dependencies:
    - "drupal/devel:1.x-dev"
    - "drupal/rules:3.x-dev"
    - "drupal/panels:4.x-dev"
    - "drupal/page_manager:1.x-dev"
   - "drupal/context:1.x-dev"
   - "drupal/layout:1.x-dev"
   - "drupal/layout_plugin:1.x-dev"
   - "drupal/bootstrap_layouts:5.x-dev"
  drupal_enable_modules: [ 'devel', 'rules', 'panels', 'page_manager'] #'context', layout', 'layout_plugin', 'bootstrap_layouts' ]

**Comment by oxyc** [comment]
That should work. Could you post your `config.yml` and what exactly the parser error says.

#1163 #1190

**Comment by laurencefass** [comment]
exact contents of my config.yml >>>

drupal_build_composer_project: false
drupal_build_composer: true
drupal_build_makefile: false
drupal_core_path: "{{ drupal_composer_install_dir }}/docroot"

drupal_composer_dependencies:
  - "drupal/devel:1.x-dev"
  - "drupal/rules:3.x-dev"
  - "drupal/panels:4.x-dev"
  - "drupal/page_manager:1.x-dev"
  - "drupal/context:1.x-dev"
  - "drupal/layout:1.x-dev"
  - "drupal/layout_plugin:1.x-dev"
  - "drupal/bootstrap_layouts:5.x-dev"
drupal_enable_modules: [ 'devel' , 'rules', 'panels', 'page_manager', 'context', layout', 'layout_plugin', 'bootstrap_layouts' ]

**Comment by laurencefass** [comment]
entire VM has stopped working so ill rebuild and repost.

**Comment by laurencefass** [comment]
my drupal.composer.json file included here. this also contains devel. do i need to require all my contrib modules here too (in addition to config.yml)?

{
    "name": "custom-project/drupal-vm",
    "description": "",
    "type": "project",
    "license": "MIT",
    "authors": [
        {
            "name": "",
            "role": ""
        }
    ],
    "repositories": [
        {
            "type": "composer",
            "url": "https://packages.drupal.org/8"
        }
    ],
    "require": {
        "composer/installers": "^1.0.20",
        "drupal-composer/drupal-scaffold": "^2.0.1",
        "drupal/core": "~8.1",
        "drupal/devel": "1.x-dev"
    },
    "replace": {
        "drupal/drupal": "*"
    },
    "minimum-stability": "dev",
    "prefer-stable": true,
    "extra": {
        "installer-paths": {
            "docroot/core": ["type:drupal-core"],
            "docroot/modules/contrib/{$name}": ["type:drupal-module"],
            "docroot/profiles/contrib/{$name}": ["type:drupal-profile"],
            "docroot/themes/contrib/{$name}": ["type:drupal-theme"],
            "drush/contrib/{$name}": ["type:drupal-drush"]
        }
    },
    "config": {
        "process-timeout": 1200
    }
}

**Comment by oxyc** [comment]
That should be done automatically with `drupal_composer_dependencies`. What's the parsing error you get?

**Comment by laurencefass** [comment]
sorry i dont understand your last comment. what should be done automatically? im assuming i need to specify contrib modules somewhere? there are two potential entry points composer.json and config.yaml. which is the correct one? or do i need to add entries to both?

**Comment by oxyc** [comment]
> my initial best guess... added these entries to my config.yml file... this throws parser errors....

I'm wondering what the parser errors you get are?

---

You can either add the dependencies to `drupal.composer.json` or list them in the `drupal_composer_dependencies` list in `config.yml`. You do not need to do both.

**Comment by laurencefass** [comment]
. You do not need to do both. ok. will try this. one problem becomes apparent: if i dont specify them in the config.yml file what will happen if i try to enable them in that file? will it blow up? is there a preferred/recommended way of adding contrib file to a project? thanks.

**Comment by laurencefass** [comment]
ok ive run again and i cant see parsing errors this time. ive removed all the drupal.composer.json entries and opted for config.yml.  it looks as though my setup is using composer as its only installing devel (specified in the composer require section and ignoring the entries in my config.yml. if this config is ignoring my config.yml entries how vagrant determine which ones to enable?

output....

**TASK [geerlingguy.drupal : Install dependencies with composer require.] ********
ok: [drupalvm] => (item=drupal/devel:1.x-dev)**

TASK [geerlingguy.drupal : include] ********************************************
skipping: [drupalvm]

TASK [geerlingguy.drupal : Check if a project specific Drush binary exists.] ***
ok: [drupalvm]

TASK [geerlingguy.drupal : Use project specific Drush if available.] ***********
skipping: [drupalvm]

TASK [geerlingguy.drupal : include] ********************************************
included: /vagrant/provisioning/roles/geerlingguy.drupal/tasks/install-site.yml for drupalvm

TASK [geerlingguy.drupal : Check if site is already installed.] ****************
ok: [drupalvm]

TASK [geerlingguy.drupal : Configure database correctly if using PostgreSQL.] **
skipping: [drupalvm]

TASK [geerlingguy.drupal : Install Drupal with drush.] *************************
skipping: [drupalvm]

TASK [geerlingguy.drupal : Install configured modules with drush.] *************
skipping: [drupalvm]

TASK [Check if local Drush configuration folder exists.] ***********************
ok: [drupalvm -> 127.0.0.1]

TASK [Create Drush configuration folder if it doesn't exist.] ******************
skipping: [drupalvm]

TASK [Configure host machine drush aliases.] ***********************************
changed: [drupalvm -> 127.0.0.1]

TASK [Ensure drush directory exists for vagrant user inside VM.] ***************
ok: [drupalvm]

TASK [Configure drush aliases for vagrant user inside VM.] *********************
changed: [drupalvm]

TASK [Ensure drush directory exists for root user inside VM.] ******************
ok: [drupalvm]

TASK [Configure drush aliases for root user inside VM.] ************************
ok: [drupalvm]

TASK [Ensure configured cron jobs exist in user account's crontab.] ************

TASK [Ensure the dashboard directory exists.] **********************************
ok: [drupalvm]

TASK [Copy dashboard page into place.] *****************************************
ok: [drupalvm]

TASK [Copy phpinfo file into place.] *******************************************
ok: [drupalvm]

TASK [Run configured post-provision shell scripts.] ****************************

TASK [Run configured post-provision ansible task files.] ***********************

PLAY RECAP *********************************************************************
drupalvm                   : ok=187  changed=4    unreachable=0    failed=0


laurence@windows-8 MINGW64 ~/VirtualBox VMs/Vagrant Boxes/drupal-vm (master)

**Comment by laurencefass** [comment]
my config.yml file looks exactly like this:

drupal_build_composer_project: false
drupal_build_composer: true
drupal_build_makefile: false
drupal_core_path: "{{ drupal_composer_install_dir }}/docroot"

drupal_composer_dependencies:
  - "drupal/devel:1.x-dev"
  - "drupal/rules:3.x-dev"
  - "drupal/panels:4.x-dev"
  - "drupal/page_manager:1.x-dev"
  - "drupal/context:1.x-dev"
  - "drupal/layout:1.x-dev"
  - "drupal/layout_plugin:1.x-dev"
  - "drupal/bootstrap_layouts:5.x-dev"
drupal_enable_modules: [ 'devel' , 'rules', 'panels', 'page_manager', 'context', layout', 'layout_plugin', 'bootstrap_layouts' ]

**Comment by geerlingguy** [comment]
Just as a note: the `drupal_composer_dependencies` variable is simply a convenience in case you're wanting to build templatized Drupal sites quickly (e.g. 'give me a fresh Drupal 8 site with these modules, and let me do it a hundred times').

If you're building one new Drupal site, and want to maintain it going forward, then it's best to manage Composer dependencies using Composer alone, outside of the Drupal VM provisioning workflow (it's more efficient that way).

See my recent posts:

  - [Using Drupal VM to build local and prod](https://www.jeffgeerling.com/blog/2017/soup-nuts-using-drupal-vm-build-local-and-prod)
  - [Tips for Managing Drupal 8 projects with Composer](https://www.jeffgeerling.com/blog/2017/tips-managing-drupal-8-projects-composer)

Basically, once you have a Drupal site, the quickest and best way to add new modules via Composer is to just run `composer require drupal/panels:4.x-dev` (as an example).

Also, the `drupal_enable_modules` modules will only be enabled during the first provisioning run, when Drupal is initially installed. After that, it's best if you either have a persistent database, or export your site's configuration and manage what modules are installed or uninstalled using configuration.

**Comment by geerlingguy** [comment]
Please let us know if there are any other questions, or if this issue can be closed. Also, if you have any failures/error messages, please copy them in their entirety and paste them here. You can use backticks  (\`) to indicate code blocks (like this: `code block`).

**Comment by laurencefass** [comment]
I am using the following settings...
drupal_build_composer_project: false
drupal_build_composer: true

I still dont really understand the difference. Can you summarise the difference so i can make a choice on future provisions. Remember I stopped using the composer_project because it was hanging on the provision... 

Will both of these options import the 'drupal_composer_dependencies' modules?

**Comment by geerlingguy** [comment]
@laurencefass -

* **drupal_build_composer_project** builds a new drupal project (with all kinds of structure, extra utilities, and helper stuff) using https://github.com/drupal-composer/drupal-project
* **drupal_build_composer** takes an existing composer.json file (like the example one included with Drupal VM, but you can use any that you make) and runs `composer install` on it.

In either case, adding more modules is a simple affair, just calling out to `composer require drupal/[modulename]:constraint`.

Either way you go, you'll end up with a Drupal site that has modules with it. One way builds a project based on a giant template managed by the people behind drupal-project. The other way just grabs Drupal core and that's about it.

But in both cases, you would use Composer to manage your site's dependencies (modules, themes, libraries, dev tools, etc.).

Drupal VM has a convenience feature that lets you build one, ten, or a thousand Drupal projects based either on a composer.json template or drupal-project. But if you're just wanting to build one new Drupal site (e.g. www.jeffgeerling.com), you might be even better off building your own codebase, and maintaining all the Composer stuff on your own, then just integrating Drupal VM into the project using Composer too. (Treat Drupal VM like another one of many development tools).

It's a little confusing, I think, because Drupal VM is flexible enough to build a Drupal site, use an existing Drupal site, integrate with multiple Drupal sites, etc.

**Comment by geerlingguy** [comment]
@laurencefass - I think it would be really helpful for you to read through the documentation on Drupal.org: [Using Composer with Drupal](https://www.drupal.org/docs/develop/using-composer/using-composer-with-drupal).

And then if you want to dig into the nitty-gritty of how the drupal-project works, read [Building better project skeletons with Composer](https://www.binpress.com/tutorial/better-project-skeletons-with-composer/157).

**Comment by geerlingguy** [comment]
Closing as there hasn't been a response in a couple weeks; please let us know if you're still having trouble.
